src/AggregateFunctions/IAggregateFunction.h

```diff
@@ -73,13 +73,19 @@ class IAggregateFunction : public std::enable_shared_from_this<IAggregateFunctio
     /// Get the data type of internal state. By default it is AggregateFunction(name(params), argument_types...).
     virtual DataTypePtr getStateType() const;
 
+    /// Same as the above but normalize state types so that variants with the same binary representation will use the same type.
```

Comment:
Why do we need non-normalized State Type?
I thought, that if the states are the same, they should use the same type.

Reply:
Because `getStateType()` is used to do correct finalization of aggregate states.


```
-- If we return original state type for quantiles*
SELECT finalizeAggregation(quantilesTimingState(0.5)(number)) FROM numbers(10)
┌─finalizeAggregation(quantilesTimingState(0.5)(number))─┐
│ [5]                                                    │
└────────────────────────────────────────────────────────┘

-- If we return normalized state type for quantiles*, which param is always 1
SELECT finalizeAggregation(quantilesTimingState(0.5)(number)) FROM numbers(10)
┌─finalizeAggregation(quantilesTimingState(0.5)(number))─┐
│ [9]                                                    │
└────────────────────────────────────────────────────────┘
```

It's because of [this](https://github.com/amosbird/ClickHouse/blob/better-projection1-fix1/src/AggregateFunctions/AggregateFunctionState.h#L37)

Other test failures if we only use normalized state types: https://s3.amazonaws.com/clickhouse-test-reports/39420/69347028c54edcedc9a43e6795c52c15ad6972ec/fast_test.html

Reply:
Ok, understood. Let's keep it with two separate functions.